homeassistant/components/ws66i/media_player.py

```diff
@@ -0,0 +1,278 @@
+    CONF_SOURCES,
+    DOMAIN,
+    FIRST_RUN,
+    SERVICE_RESTORE,
+    SERVICE_SNAPSHOT,
+    WS66I_OBJECT,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PARALLEL_UPDATES = 1
+
+SUPPORT_WS66I = (
+    SUPPORT_VOLUME_MUTE
+    | SUPPORT_VOLUME_SET
+    | SUPPORT_VOLUME_STEP
+    | SUPPORT_TURN_ON
+    | SUPPORT_TURN_OFF
+    | SUPPORT_SELECT_SOURCE
+)
+
+
+@core.callback
+def _get_sources_from_dict(data):
+    sources_config = data[CONF_SOURCES]
+
+    source_id_name = {int(index): name for index, name in sources_config.items()}
+
+    source_name_id = {v: k for k, v in source_id_name.items()}
+
+    source_names = sorted(source_name_id.keys(), key=lambda v: source_name_id[v])
+
+    return [source_id_name, source_name_id, source_names]
+
+
+@core.callback
+def _get_sources(config_entry):
+    if CONF_SOURCES in config_entry.options:
+        data = config_entry.options
+    else:
+        data = config_entry.data
+    return _get_sources_from_dict(data)
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the WS66i 6-zone amplifier platform."""
+    addr = config_entry.data[CONF_IP_ADDRESS]
+
+    # Open up a connection with the WS66i
+    ws66i = hass.data[DOMAIN][config_entry.entry_id][WS66I_OBJECT]
+    try:
+        await hass.async_add_executor_job(ws66i.open)
+    except ConnectionError as err:
+        raise ConfigEntryNotReady from err
+
+    sources = _get_sources(config_entry)
+
+    # Build the entities that control each of the zone.
+    entities = []
+    for i in range(1, 4):
+        for j in range(1, 7):
+            zone_id = (i * 10) + j
+            _LOGGER.info("Adding zone %d for address %s", zone_id, addr)
+            entities.append(Ws66iZone(ws66i, sources, config_entry.entry_id, zone_id))
+
+    # only call update before add if it's the first run so we can try to detect zones
+    first_run = hass.data[DOMAIN][config_entry.entry_id][FIRST_RUN]
+    async_add_entities(entities, first_run)
+
+    def close(event):
+        """Close the Telnet connection."""
+        ws66i.close()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, close)
+
+    platform = entity_platform.async_get_current_platform()
+
+    def _call_service(entities, service_call):
+        for entity in entities:
+            if service_call.service == SERVICE_SNAPSHOT:
+                entity.snapshot()
+            elif service_call.service == SERVICE_RESTORE:
+                entity.restore()
+
+    @service.verify_domain_control(hass, DOMAIN)
+    async def async_service_handle(service_call):
+        """Handle for services."""
+        entities = await platform.async_extract_from_service(service_call)
+
+        if not entities:
+            return
+
+        hass.async_add_executor_job(_call_service, entities, service_call)
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_SNAPSHOT,
+        async_service_handle,
+        schema=cv.make_entity_service_schema({}),
```

Comment:
```suggestion
        schema={},
```

Reply:
Use entity services and pass a `dict` instead

https://developers.home-assistant.io/docs/dev_101_services?_highlight=entity&_highlight=services#entity-services